### PR TITLE
app: support more hashes for X-Hub-Signature

### PR DESCRIPTION
**Pull request by robbat2**
Signed-off-by: Robin H. Johnson <robbat2@gentoo.org>

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -17,6 +17,8 @@
     'DEBUG': os.environ.get('DEBUG', '') in [1, '1', 'True', 'true'],
     'SECRET': os.environ.get('SECRET'),
     'S3_REGION': os.environ.get('S3_REGION', 'eu-west-1'),
+    'HASHLIB_BLACKLIST': map(lambda s: s.strip().lower(), \
```
Comment (codingjoe) [inline app.py]: This will return a generator. I would recommend to cast this into a set.

File: app.py [hunk 2]
```diff
@@ -31,14 +33,17 @@ def validate_signature(request):
...
         signature = request.headers['X-Hub-Signature']
-        _, sha1 = signature.split('=')
+        hashname, hashval = signature.split('=')
     except (KeyError, ValueError):
         raise BadRequestError()
-    digest = hmac.new(CONFIG['SECRET'], request.raw_body, hashlib.sha1) \
-        .hexdigest()
-    if not hmac.compare_digest(digest, sha1.encode('utf-8')):
-        raise UnauthorizedError()
+    if (hashname in CONFIG['HASHLIB_BLACKLIST']) or \
```
Comment (codingjoe) [inline app.py:47]: If the config is a set this would be helpful here too, to increase readability.

File: app.py [hunk 3]
```diff
@@ -31,14 +33,17 @@ def validate_signature(request):
...
+            (hashname not in hashlib.algorithms_available):
+        raise UnauthorizedError('X-Hub-Signature hash unavailable')
```
Comment (codingjoe) [inline app.py]: I would raise a 400 (not 403) if the signature is not the expected hash algo.
Also I would raise two different errors if the algo is blacklisted or if it is not available.

**Review by codingjoe — COMMENTED** [review]
HI @robbat2 not a bad idea. I don't know if I wouldn't prefer a whitelist, but if GitHub ever decides to change it without notice, this would be bad.

That being said, cool PR. Only thing missing really is a test ;)

**Comment by codingjoe** [comment]
/ping @robbat2
do you still want to finish this?

**Comment by robbat2** [comment]
Yes, let me refactor from my latest local version. Getting it deployed at the time was more important than coming back to get it upstream.

**Comment by codingjoe** [comment]
ok, can't wait 👍

**Comment by codingjoe** [comment]
@robbat2 are you still with working on this? I'd love to get it merged, but it needs some final touches and a rebase.

**Comment by robbat2** [comment]
@codingjoe as promised, now rebased

**Comment by robbat2** [comment]
@codingjoe now with tests as well; Not sure how to make @codecov re-run the coverage suite.

**Comment by codingjoe** [comment]
@robbat2 I did some minor changes and merged.

**Pull request merged**